// frosted-ui/dist/cjs/components/avatar-stack/avatar-stack.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Size = exports.Default = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const __1 = require("..");
// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
    title: 'Components/AvatarStack',
    component: __1.AvatarStack.Root,
    parameters: {
        // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
        layout: 'centered',
    },
    // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
    tags: ['autodocs'],
};
exports.default = meta;
const availableUsers = [
    {
        name: 'Artur Bień',
        color: 'blue',
        src: 'https://images.unsplash.com/photo-1502823403499-6ccfcf4fb453?&w=256&h=256&q=70&crop=focalpoint&fp-x=0.5&fp-y=0.3&fp-z=1&fit=crop',
    },
    {
        name: 'Cameron Zoub',
        color: 'green',
        src: '',
    },
    {
        name: 'Carlos Whitt',
        color: 'pink',
        src: '',
    },
    {
        name: 'Ilya Miskov',
        color: 'yellow',
        src: '',
    },
    {
        name: 'Guillermo',
        color: 'bronze',
        src: '',
    },
    {
        name: 'Steven Schwartz',
        color: 'lime',
        src: '',
    },
    {
        name: 'FJ',
        color: 'tomato',
        src: '',
    },
];
// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
exports.Default = {
    args: {},
    render: (args) => {
        return (react_1.default.createElement("div", { style: { display: 'flex', flexDirection: 'column', gap: 'var(--space-5)' } },
            react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "3" }, availableUsers.map((user, index) => (react_1.default.createElement(__1.AvatarStack.Avatar, { shape: index % 2 === 0 ? 'circle' : 'square', fallback: user.name, color: user.color, key: user.name, src: user.src })))),
            react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "3" }, availableUsers.map((user) => (react_1.default.createElement(__1.AvatarStack.Avatar, { shape: "square", fallback: user.name, color: user.color, key: user.name, src: user.src })))),
            react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "3" },
                react_1.default.createElement(__1.AvatarStack.Avatar, { fallback: availableUsers[0].name, color: availableUsers[0].color, src: availableUsers[0].src }))));
    },
};
exports.Size = {
    args: {},
    render: (args) => (react_1.default.createElement("div", { style: { display: 'flex', flexDirection: 'column', gap: 'var(--space-4)' } },
        react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "0" }, availableUsers.map((user) => (react_1.default.createElement(__1.AvatarStack.Avatar, { fallback: user.name, color: user.color, key: user.name, src: user.src })))),
        react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "1" }, availableUsers.map((user) => (react_1.default.createElement(__1.AvatarStack.Avatar, { fallback: user.name, color: user.color, key: user.name, src: user.src })))),
        react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "2" }, availableUsers.map((user) => (react_1.default.createElement(__1.AvatarStack.Avatar, { fallback: user.name, color: user.color, key: user.name, src: user.src })))),
        react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "3" }, availableUsers.map((user) => (react_1.default.createElement(__1.AvatarStack.Avatar, { fallback: user.name, color: user.color, key: user.name, src: user.src })))),
        react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "4" }, availableUsers.map((user) => (react_1.default.createElement(__1.AvatarStack.Avatar, { fallback: user.name, color: user.color, key: user.name, src: user.src })))),
        react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "5" }, availableUsers.map((user) => (react_1.default.createElement(__1.AvatarStack.Avatar, { fallback: user.name, color: user.color, key: user.name, src: user.src })))),
        react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "6" }, availableUsers.map((user) => (react_1.default.createElement(__1.AvatarStack.Avatar, { fallback: user.name, color: user.color, key: user.name, src: user.src })))),
        react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "7" }, availableUsers.map((user) => (react_1.default.createElement(__1.AvatarStack.Avatar, { fallback: user.name, color: user.color, key: user.name, src: user.src })))),
        react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "8" }, availableUsers.map((user) => (react_1.default.createElement(__1.AvatarStack.Avatar, { fallback: user.name, color: user.color, key: user.name, src: user.src })))),
        react_1.default.createElement(__1.AvatarStack.Root, { ...args, size: "9" }, availableUsers.map((user) => (react_1.default.createElement(__1.AvatarStack.Avatar, { shape: 'square', fallback: user.name, color: user.color, key: user.name, src: user.src })))))),
};
