// @whop-apps/dev-proxy/dist/parse-cli-options.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "getRunOptions", {
    enumerable: true,
    get: function() {
        return getRunOptions;
    }
});
const _nodeutil = require("node:util");
function getRunOptions() {
    const { values } = (0, _nodeutil.parseArgs)({
        options: {
            proxyPort: {
                type: "string"
            },
            upstreamPort: {
                type: "string"
            },
            npmCommand: {
                type: "string"
            },
            command: {
                type: "string"
            },
            appComponentOrigins: {
                type: "string"
            },
            standalone: {
                type: "boolean",
                default: false
            }
        }
    });
    return {
        ...values,
        proxyPort: values.proxyPort ? Number(values.proxyPort) : undefined,
        upstreamPort: values.upstreamPort ? Number(values.upstreamPort) : undefined,
        appComponentOrigins: values.appComponentOrigins ? JSON.parse(values.appComponentOrigins) : undefined
    };
}
